rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,526 @@
+the status endpoint.
+
+We may also want to have the `/status` endpoint display each individual worker status in addition to the 
+global status, which may be breaking change in the `/status` API response format.
+
+### PID file
+
+Without changes, each worker is going to try to write and read the same PID file. Also, this breaks the whole pid file 
+usage, as the PID stored in the file will be a arbitrary worker’s PID, instead of the coordinator (main process) PID.
+
+In clustering mode, we will need to have to coordinator handle the PID file logic, and to disable pid file handling 
+in the worker's environment service.
+
+### SavedObjects migration
+
+In the current state, all workers are going to try to perform the migration. Ideally, we would have only one process 
+perform the migration, and the other ones just wait for a ready signal. We can’t easily have the coordinator do it, 
+so we would probably have to leverage the ‘main worker’ concept here.
+
+The SO migration v2 is supposed to be resilient to concurrent attempts though, as we already support multi-instances 
+Kibana, so this can probably be considered an improvement.
+
+## Open questions and things to solve
+
+### Memory consumption
+
+In clustered mode, node options such as `max-old-space-size` will be used by all processes. 
+
+E.g using `--max-old-space-size=1024` in a 2 workers cluster would have a maximum memory usage of 3gb (1 coordinator + 2 workers). 
+
+If this something we will need to document somewhere?
+
+### Workers error handling
+
+When using `cluster`, the common best practice is to have the coordinator recreate ('restart') workers when they terminate unexpectedly. 
+However, given Kibana's architecture, some failures are not recoverable (workers failing because of config validation, failed migration...). 
+
+For instance, if a worker (well, all workers) terminates because of an invalid configuration property, it doesn't make
+any sense to have the coordinator recreate them indefinitely, as the error requires manual intervention. 
+
+Should we try to distinguish recoverable and non-recoverable errors, or are we good terminating the main Kibana process 
+when any worker terminates unexpectedly for any reason (After all, this is already the behavior in non-cluster mode)?
+
+### Data folder
+
+The data folder (`path.data`) is currently the same for all workers. 
+
+We still have to identify with the teams if this is going to be a problem. It could be, for example, if some plugins
+are accessing files in write mode, which could result in concurrency issues between the workers.
+
+If that was confirmed, we would have to create and use a distinct data folder for each worker.
+
+One pragmatic solution could be, when clustering is enabled, to create a sub folder under path.data for each worker. 
+
+If we need to do so, should this be considered a breaking change, or do we assume the usage of the data folder is an 
+implementation detail and not part of our public 'API'?
```

Comment:
I don't believe we consider this part of our public API and I'm pretty sure we've changed this in the past during minor releases.